### ISSUE TITLE
VoiceOver: Unable to interact with appointments if swipe before appointments finish loading  [Lvl 1]

### ISSUE THREAD

**Issue by outerpress** [department-of-veterans-affairs/va.gov-team#111648]
## Description

After the page loads and I swipe right at least once *while* appointments are still loading (for both upcoming and past appointments) I cannot fully interact with the appointments list. This could very well happen for users reading other content on the screen while waiting for their appointments to load; it may cause confusion or worry if they are unable to access their appointments.  

 \[Insert recording into GitHub ticket\]  

* **Page:** Appointments (Upcoming and Past)  
* **Device:** iPhone  
* **Orientation:** Both  
* **Expected severity level:** 1  
* **Expected experience standard:** User can complete necessary flows  
* **Test user:** Judy  
* **How to recreate:** Navigate to either Upcoming or Past appointments while using VoiceOver in Safari and swipe at least once while the appointments are loading.  
* **Recommendation:** I recommend doing a technical spike to determine why focus management is breaking after the appointments list loads. A quicker solution may be to consider having an entire page loading indicator so Veterans can’t interact with other elements on the page before the appointments finish loading.

---

## Development Checklist
Background: the `toggle` toggle is on (remove if not needed) 

- [ ] Task 1
- [ ] Task 2

### Design Assets
- [Figma file]() (put N/A if not needed) 

### Developer Reference

- [Swagger File]() (remove if not needed)
- [File path]() (remove if not needed)

## Definition of Done
- [ ] All tasks criteria are met
- [ ] [Technical Documentation]() is updated (must confirm if needed/or remove) 
- [ ] [Feature Reference (GitHub)](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/health-care/appointments/va-online-scheduling/feature-reference) is updated
- [ ] UX - [Feature Reference (Figma)](https://www.figma.com/design/eonNJsp57eqfPqx7ydsJY9/Feature-Reference-%7C-Appointments-FE?node-id=1150-105402&t=zRNPBqgPDVU0dDP8-1) is updated

---

**Comment by outerpress** [comment]
@coforma-jamie @BobbyBaileyRB do you have a recording of this one? I didn't see one in the original ticket, and I don't have an iPhone to recreate.

Also, does ADE offer testing support if a team doesn't have access to the right devices/tools?

Thanks! FYA @jamiefiore

**Comment by coforma-jamie** [comment]
That's the new version in beta? Curious if the issue will still be fixed in iOS 26 by the time it releases to the general public.

For what it's worth, I was able to duplicate the error using an iPhone 14 on iOS 18.5.

**Comment by coforma-jamie** [comment]
Confirmed that this is also happening on iPadOS 18.5. (Having a tech issue so can't grab a recording unfortunately.)

**Comment by eselkin** [comment]
Ok, @outerpress @coforma-jamie I reproduced the issue on an iPad (I bought one because the VA said iPhones are on backorder. I asked for a used one so they should get that to me Friday hopefully)... It is interesting. You can get to the appointment list on screen but swiping left/right does not bring you to the links... 

If you click into the section that has the appointments list then you can swipe to get to the links or items, but you definitely shouldn't have to do that, because if you can't see it, you wouldn't know that section was there at all. There must be like a tabindex=-1 or something on the div. I'll investigate. 

@coforma-jamie What do you use to get screen recordings on ipad or iphones?

**Comment by coforma-jamie** [comment]
There's a built-in screen recorder app :) Info - https://support.apple.com/en-us/102653

**Comment by simiadebowale** [comment]
This ticket is currently blocked by - https://github.com/orgs/department-of-veterans-affairs/projects/1554/views/6?pane=issue&itemId=123005670&issue=department-of-veterans-affairs%7Cva.gov-team%7C116010

**Comment by eselkin** [comment]
So I've tried
1. Adding solely a placeholder div (no change to access appointments in the list for VO on iOS/iPad)
2. Adding a placeholder div and aria-busy to an underlying div (no change)
3. Adding the loading indicator inside the div with the aria-busy updating (no change)
4. Setting the loading indicator to always have focus (with set-focus) to see if losing focus tells VO to look for changes (no change)
5. Adding ids to list elements (no change)
6. Adding aria-atomic and aria-busy to the div that gets updated with the appointments list (no change)
7. Adding aria-atomic to individual list items (because they call a function - no change)
8. Adding extra roles to the list items and lists (no change, not expected to work)
9. Adding aria-live="polite" **does** announce the items when they get added... which we **don't want to do** (but it's interesting that VoiceOver knows they are there though)
10. Adding role="status" **does** announce the items... which we **don't want to do** (same as above)Interestingly,
If you scroll up in the page (3 finger scroll up or down), the list becomes available by one-finger selecting (starting from anywhere on the page).

**In conclusion**, unless we add the focus to the initial element of the dynamic list then none of the list elements (except randomly the last one) are accessible.

We can't do certain things on the page because of the SPA nature, like dynamically alter the text of the list elements with a element.textContent (that is often used to make VoiceOver notice changes). We can't announce the elements.

It may have something to do with this page's list complexity (nested lists in lists -- similar to the calendar list of lists) that makes it not easy on screen readers (particularly VO). I've noted that we should think twice before using complex dynamic nested list structures because of this. However, it may just be a shortcoming of VoiceOver on mobile devices and react apps like vets-website.

**Comment by eselkin** [comment]
In an attempt to see what the issue was I found out something interesting about VoiceOver on mobile devices.

If you have dynamic content and scroll past that content (whether it has a loading indicator or not) VoiceOver no longer picks up that dynamic (list) content. It can be any type of content evidently. Even if you change the ids/keys/names after it loads! So basically, anything dynamic can suffer this issue.

https://github.com/user-attachments/assets/b187e683-25d2-4bbc-823d-ff7be07eb3af

**Comment by outerpress** [comment]
Adding a note for the future overall enhancement work (not in scope for this ticket) - @eselkin and I were discussing ways to mitigate this issue by loading the list more promptly:
- Load the list earlier, possibly, during the check to see if the page/app is working. 
- Make the whole page wait for the list to load before showing anything. 
- Lazy-load the list (@simiadebowale you brought this up awhile back)
- Add pagination

**Comment by coforma-jamie** [comment]
This all makes sense to me @outerpress and @eselkin. Thank you for doing such a deep dive, and thanks to @amyleadem for looking in to this too.
